### PR TITLE
Add Mnist Data.

### PR DESCRIPTION
**Pull request by rilmayer**
I tried to write code about Mnist(http://yann.lecun.com/exdb/mnist/).
I think we will be able to use Mnist Dataset.

### CODE REVIEW

File: lib/datasets/mnist.rb [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        end
+      end
+      image_arrays
+    end
+
+    def parse_labels(file_path)
+      labels = []
+      Zlib::GzipReader.open(file_path) do |f|
+        magic, n_labels = f.read(8).unpack('N2')
+        labels = f.read(n_labels).unpack('C*')
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can we remove the `labels` local variable here?

For example:

```ruby
Zlib::GzipReader.open do |f|
  # ...
  f.read(...)...
end
```

File: test/test-mnist.rb [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    @dataset.cache_dir_path = @tmp_dir
+
+    def @dataset.download(output_path, url)
+
+      directory = (Pathname.new(__dir__) + "tmp").expand_path
+      image_paths = ["#{directory}/train-images-idx3-ubyte.gz",
+                     "#{directory}/t10k-images-idx3-ubyte.gz"]
+      label_paths = ["#{directory}/train-labels-idx1-ubyte.gz",
+                     "#{directory}/t10k-labels-idx1-ubyte.gz"]
+      image_num, image_size_x, image_size_y, label = 10, 28, 28, 1
```
Comment (kou) [inline test/test-mnist.rb]: `n_images` is better because we use `n_XXX` for the number of XXX.

File: test/test-mnist.rb [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+      Zlib::GzipWriter.open(output_path) do |gz|
+        if image_paths.include?(output_path.to_s)
```
Comment (kou) [inline test/test-mnist.rb]: Can we use `output_path.basename.to_s.include?("-images-")` here?
It seems that `Pathname.new(__dir__) + "tmp"` is too implementation specific.

File: lib/datasets/mnist.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      @metadata.url = "http://yann.lecun.com/exdb/mnist/"
+      @type = type
+
+      case type
+      when :train
+        @metadata.description = "a training set of 60,000 examples"
+      when :test
+        @metadata.description = "a test set of 10,000 examples"
+      end
+
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can you remove an empty before `end`?

File: README.md [hunk 5]
```diff
@@ -73,6 +73,24 @@ cifar.each do |record|
 end
 ```
 
+**Mnist**
```
[the remark this answers is not in the document]
Comment (rilmayer) [reply README.md]: I think your idea "MNIST" is better than using "Mnist".
So I fixed it.

File: lib/datasets/mnist.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      images = parse_images(image_path)
+      lables = parse_labels(label_path)
+      images.zip(lables) do |row|
+        pixels = row[0]
+        label = row[1]
+        yield Record.new(pixels, label)
+      end
+    end
+
+    def target_file(data: :img)
```
[the remark this answers is not in the document]
Comment (rilmayer) [reply lib/datasets/mnist.rb]: OK, I fixed it.

File: lib/datasets/mnist.rb [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      end
+
+    end
+
+    def each(&block)
+      return to_enum(__method__) unless block_given?
+
+      image_path = cache_dir_path + target_file(data: :img)
+      label_path = cache_dir_path + target_file(data: :label)
+      unless image_path.exist? && label_path.exist?
```
[the remark this answers is not in the document]
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I stopped using `unless` and `&&` combination.

File: lib/datasets/mnist.rb [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+          "t10k-images-idx3-ubyte.gz"
+        when :label
+          "t10k-labels-idx1-ubyte.gz"
+        end
+      end
+    end
+
+    def parse_images(file_path)
+      image_arrays = []
+      n_rows ,n_cols = 0, 0
```
[the remark this answers is not in the document]
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I fixed it.

File: lib/datasets/mnist.rb [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      Zlib::GzipReader.open(file_path) do |f|
+        magic, n_images = f.read(8).unpack('N2')
+        n_rows, n_cols = f.read(8).unpack('N2')
+        n_images.times do
+          img_str = f.read(n_rows * n_cols)
+          # IMAGE SIZE is 784
+          image_arrays << img_str.each_byte.to_a.each_slice(784).to_a[0]
```
[the remark this answers is not in the document]
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I think your suggestion is very good, so I fixed it.

File: lib/datasets/mnist.rb [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      end
+    end
+
+    def parse_images(file_path)
+      image_arrays = []
+      n_rows ,n_cols = 0, 0
+      Zlib::GzipReader.open(file_path) do |f|
+        magic, n_images = f.read(8).unpack('N2')
+        n_rows, n_cols = f.read(8).unpack('N2')
+        n_images.times do
```
[the remark this answers is not in the document]
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I adopted your suggestion.

File: lib/datasets/mnist.rb (hunk 1 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I removed `labels`.
Thanks to your advice, the code improved a lot.

File: test/test-mnist.rb (hunk 2 above)
Comment (rilmayer) [reply test/test-mnist.rb]: I fixed it.

File: test/test-mnist.rb (hunk 3 above)
Comment (rilmayer) [reply test/test-mnist.rb]: I understood what you say, so I fixed it.

File: lib/datasets/mnist.rb (hunk 4 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I fixed it.

File: lib/datasets/mnist.rb [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+      end
+
+      open_data(image_path, label_path, &block)
+    end
+
+    private
+    def open_data(image_path, label_path, &block)
+      images = parse_images(image_path)
+      lables = parse_labels(label_path)
+      images.zip(lables) do |row|
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can we implement stream parsing instead of "parsing then yielding records"?

File: lib/datasets/mnist.rb [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    end
+
+    def parse_labels(file_path)
+      Zlib::GzipReader.open(file_path) do |f|
+        magic, n_labels = f.read(8).unpack('N2')
+        f.read(n_labels).unpack('C*')
+      end
+    end
+  end
+end
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can you add a newline at the end of file?

File: test/test-mnist.rb [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+      end
+    end
+  end
+
+  def teardown
+    teardown_sandbox
+  end
+
+  sub_test_case("mnist") do
+
```
Comment (kou) [inline test/test-mnist.rb]: Can you remove an empty line after `do`?

File: lib/datasets/mnist.rb [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        when :img
+          "t10k-images-idx3-ubyte.gz"
+        when :label
+          "t10k-labels-idx1-ubyte.gz"
+        end
+      end
+    end
+
+    def parse_images(file_path)
+      n_rows, n_cols = 0, 0
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can we remove them?

File: lib/datasets/mnist.rb [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+      Zlib::GzipReader.open(file_path) do |f|
+        magic, n_images = f.read(8).unpack('N2')
+        n_rows, n_cols = f.read(8).unpack('N2')
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can we remove the `8` magic number?

```ruby
n_uint32s = 4
n_bytes = n_uint32s * 4
magic, n_images, n_rows, n_cols = f.read(n_bytes).unpack("N*")
```

File: lib/datasets.rb [hunk 16]
```diff
@@ -3,3 +3,4 @@
 require "datasets/cifar"
 require "datasets/iris"
 require "datasets/wikipedia"
+require "datasets/mnist"
```
Comment (kou) [inline lib/datasets.rb]: Can you sort alphabetically?

File: lib/datasets/mnist.rb [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        when :label
+          "t10k-labels-idx1-ubyte.gz"
+        end
+      end
+    end
+
+    def parse_images(file_path)
+      n_rows, n_cols = 0, 0
+      Zlib::GzipReader.open(file_path) do |f|
+        magic, n_images = f.read(8).unpack('N2')
```
Comment (kou) [inline lib/datasets/mnist.rb]: Should we validate the magic?

File: lib/datasets/mnist.rb [hunk 18]
```diff
@@ -0,0 +1,96 @@
+require 'zlib'
+
+require_relative "dataset"
+
+
+module Datasets
+  class Mnist < Dataset
```
[the remark this answers is not in the document]
Comment (rilmayer) [reply lib/datasets/mnist.rb]: Noted with thanks.
I have changed to use `MNIST` instead of `Mnist` .

File: lib/datasets.rb (hunk 16 above)
Comment (rilmayer) [reply lib/datasets.rb]: I fixed.

File: lib/datasets/mnist.rb (hunk 12 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I added newline.

File: lib/datasets/mnist.rb [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+      end
+    end
+
+    def parse_images(file_path)
+      n_rows, n_cols = 0, 0
+      Zlib::GzipReader.open(file_path) do |f|
+        magic, n_images = f.read(8).unpack('N2')
+        n_rows, n_cols = f.read(8).unpack('N2')
+        n_images.times.collect do
+          f.read(n_rows * n_cols).unpack('C*')
```
[the remark this answers is not in the document]
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I couldn't have thought about efficiency!
I tried to fix it by referring to CIFAR code.

File: test/test-mnist.rb (hunk 13 above)
Comment (rilmayer) [reply test/test-mnist.rb]: I removed empty line.

File: lib/datasets/mnist.rb (hunk 11 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I implemented stream parsing to only `Image data` but not to `Label data`, because I think label data is not so big that it need stream parsing.
If you think it need  to `Label data` too, for reasons such as readability, efficiency etc... , I'll fixed it!

File: lib/datasets/mnist.rb (hunk 14 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I removed them.

File: lib/datasets/mnist.rb (hunk 17 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I fixed to validate the magic.

File: lib/datasets/mnist.rb (hunk 11 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I noticed it now, we may not able to implement stream parsing to `Label data` ...

**Comment by rilmayer** [comment]
Sorry for some minor fixes, but I think I could have fixed all where you commented.

File: test/test-mnist.rb [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+            :pixels => record.pixels
+          }
+        end
+
+        assert_equal([
+                       {
+                         :label => 1,
+                         :pixels => [0] * 28 * 28
+                       }
+                      ] * 10,
```
Comment (kou) [inline test/test-mnist.rb]: It seems that there is a needless space.

File: test/test-mnist.rb [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
```
Comment (kou) [inline test/test-mnist.rb]: It seems that there is a needles space.

File: lib/datasets/mnist.rb [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+    private
+    def open_data(image_path, label_path, &block)
+      labels = parse_labels(label_path)
+
+      Zlib::GzipReader.open(image_path) do |f|
+        n_uint32s = 4
+        n_bytes = n_uint32s * 4
+        magic, n_images, n_rows, n_cols = f.read(n_bytes).unpack("N*")
+        raise 'This is not MNIST image file' if magic != 2051
+        n_images.times.collect do |i|
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can we remove `.collect` to reduce needless objects?

File: lib/datasets/mnist.rb [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+
+      def to_h
+        hash = super
+        hash[:pixels] = pixels
+        hash
+      end
+    end
+
+    class Record < Struct.new(:data, :label)
+      include Pixelable
```
Comment (kou) [inline lib/datasets/mnist.rb]: We don't need to create `Pixelable` module like CIFAR dataset.
Because there is only one record class for MNIST dataset.

File: lib/datasets/mnist.rb [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+    end
+
+    def initialize(type: :train)
+      unless [:train, :test].include?(type)
+        raise 'Please set type :train or :test'
```
Comment (kou) [inline lib/datasets/mnist.rb]: `raise ArgumentError, "type must be :train or :test: #{type.inspect}` is better.
It uses suitable exception class and has the actual value.

File: lib/datasets/mnist.rb [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+        @metadata.description = "a training set of 60,000 examples"
+      when :test
+        @metadata.description = "a test set of 10,000 examples"
+      end
+    end
+
+    def each(&block)
+      return to_enum(__method__) unless block_given?
+
+      image_path = cache_dir_path + target_file(:img)
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can we use `:image` instead of abbreviated `:img`?

File: lib/datasets/mnist.rb (hunk 24 above)
Comment (kou) [reply lib/datasets/mnist.rb]: (We should improve CIFAR dataset later.)

File: lib/datasets/mnist.rb [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+
+    private
+    def open_data(image_path, label_path, &block)
+      labels = parse_labels(label_path)
+
+      Zlib::GzipReader.open(image_path) do |f|
+        n_uint32s = 4
+        n_bytes = n_uint32s * 4
+        magic, n_images, n_rows, n_cols = f.read(n_bytes).unpack("N*")
+        raise 'This is not MNIST image file' if magic != 2051
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can you name `2051`?

File: lib/datasets/mnist.rb [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+        end
+      end
+    end
+
+    def parse_labels(file_path)
+      Zlib::GzipReader.open(file_path) do |f|
+        n_uint32s = 4
+        n_bytes = n_uint32s * 2
+        magic, n_labels = f.read(n_bytes).unpack('N2')
+        raise 'This is not MNIST label file' if magic != 2049
```
Comment (kou) [inline lib/datasets/mnist.rb]: Can you name `2049`?

File: lib/datasets/mnist.rb (hunk 26 above)
Comment (kou) [reply lib/datasets/mnist.rb]: We should have an exception class for parse error.
We should work on it later.

File: test/test-mnist.rb (hunk 20 above)
Comment (rilmayer) [reply test/test-mnist.rb]: I did not notice it at all!
I fixed.

File: test/test-mnist.rb (hunk 21 above)
Comment (rilmayer) [reply test/test-mnist.rb]: I fixed.

File: lib/datasets/mnist.rb (hunk 22 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: Thank you for pointing out `.collect` is left.
I removed `.collect`.

File: lib/datasets/mnist.rb (hunk 23 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I understood your comment.
I removed `Pixelable` module.

File: lib/datasets/mnist.rb (hunk 24 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: Thanks for your teaching! I did not know `.inspect` method, It's really useful!
I fixed.

File: lib/datasets/mnist.rb (hunk 26 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I fixed naming `2051`.
I didn't make exception class for parse error, because the class seems like util class.
So I think it is better that work on later as another pull request.

File: lib/datasets/mnist.rb (hunk 27 above)
Comment (rilmayer) [reply lib/datasets/mnist.rb]: I fixed naming `2049`.

**Comment by rilmayer** [comment]
Thanks, I have fixed the code you commented.
Please check my comments for some fix.

**Comment by kou** [comment]
Thanks!
Now, we can use it in Red Chainer for MNIST dataset.

We can work on exception classes as another work.

**Comment by rilmayer** [comment]
Thank you for having reviewed, and for having merged.
I have learned a lot, so I want to commit more.

And Thanks for some fix to my code.

**Comment by kou** [comment]
Great! We should support more datasets!
For example, we can find some target datasets at https://github.com/red-data-tools/red-datasets/issues .
